pkg/ifuzz/powerpc/powerpc.go

```diff
@@ -85,13 +119,22 @@ func (insn Insn) Encode(cfg *iset.Config, r *rand.Rand) []byte {
 	}
 
 	ret := make([]byte, 0)
-	insn32 := insn.Opcode
+	ret = append(ret, insn.encodeOpcode(cfg, r, insn.Opcode, insn.Mask, &insn.Fields)...)
+	if insn.isPrefixed() {
+		ret = append(ret, insn.encodeOpcode(cfg, r, insn.Opcode2, insn.Mask2, &insn.Fields2)...)
+	}
+	return ret
+}
+
+func (insn Insn) encodeOpcode(cfg *iset.Config, r *rand.Rand, opcode, mask uint32, f *map[string][]InsnBits) []byte {
+	ret := make([]byte, 0)
+	insn32 := opcode
 	if len(cfg.MemRegions) != 0 {
 		// The PowerISA pdf parser could have missed some fields,
 		// randomize them there.
-		insn32 |= r.Uint32() & ^insn.Mask
+		insn32 |= r.Uint32() & ^mask
 	}
-	for reg, bits := range insn.Fields {
+	for reg, bits := range *f {
```

Comment:
Not directly related to this PR, but it just caught my eye. I suspect that there's some chance for a non-deterministic behavior, as maps in Go do not give any key ordering guarantees. Earlier we already noticed that our tests reported some non-deterministic behavior, and actually powerpc instructions were a part of those differing programs.

https://github.com/google/syzkaller/issues/2790

Reply:
(after reading about maps in go) what is the solution, replace maps with array? I can do that unless there is an easier way but it looks like I'd have to generate temporary sorted slices and it is just not worth it.

Reply:
I don't mind either way. Generating temporary sorted slices seems indeed like an overkill, but it's already done in so many places in syzkaller, that one more wouldn't make a big difference. If you think that using arrays/slices is nicer in this case and it's not too much work, then it's also fine.